src/LanguageServer/Impl/CodeActions/MissingImportCodeActionProvider.cs

```diff
@@ -0,0 +1,731 @@
+using Range = Microsoft.Python.Core.Text.Range;
+
+namespace Microsoft.Python.LanguageServer.CodeActions {
+    internal sealed class MissingImportCodeActionProvider : ICodeActionProvider {
+        public static readonly ICodeActionProvider Instance = new MissingImportCodeActionProvider();
+
+        // right now, it is a static. in future, we might consider giving an option to users to customize this list
+        // also, right now, it is text based. so if module has same name, they will get same suggestion even if
+        // the module is not something user expected
+        private static readonly Dictionary<string, string> WellKnownAbbreviationMap = new Dictionary<string, string>() {
+            { "numpy", "np" },
+            { "pandas", "pd" },
+            { "tensorflow", "tf" },
+            { "matplotlib.pyplot", "plt" },
+            { "matplotlib", "mpl" },
+            { "math", "m" },
+            { "scipy.io", "spio" },
+            { "scipy", "sp" },
+        };
+
+        private MissingImportCodeActionProvider() {
+        }
+
+        public ImmutableArray<string> FixableDiagnostics => ImmutableArray<string>.Create(
+            ErrorCodes.UndefinedVariable, ErrorCodes.VariableNotDefinedGlobally, ErrorCodes.VariableNotDefinedNonLocal);
+
+        public async Task<IEnumerable<CodeAction>> GetCodeActionsAsync(IDocumentAnalysis analysis, DiagnosticsEntry diagnostic, CancellationToken cancellationToken) {
+            var finder = new ExpressionFinder(analysis.Ast, new FindExpressionOptions() { Names = true });
+            var node = finder.GetExpression(diagnostic.SourceSpan);
+            if (!(node is NameExpression nex)) {
+                return Enumerable.Empty<CodeAction>();
+            }
+
+            var identifier = nex.Name;
+            if (string.IsNullOrEmpty(identifier)) {
+                return Enumerable.Empty<CodeAction>();
+            }
+
+            var codeActions = new List<CodeAction>();
+            var diagnostics = new[] { diagnostic.ToDiagnostic() };
+
+            // see whether it is one of abbreviation we specialize
+            foreach (var moduleFullName in WellKnownAbbreviationMap.Where(kv => kv.Value == identifier).Select(kv => kv.Key)) {
+                var moduleName = GetModuleName(moduleFullName);
+
+                await GetCodeActionsAsync(analysis, diagnostics, new Input(node, moduleName, moduleFullName), codeActions, cancellationToken);
+            }
+
+            // add then search given name as it is
+            await GetCodeActionsAsync(analysis, diagnostics, new Input(node, identifier), codeActions, cancellationToken);
+
+            return codeActions;
+
+            string GetModuleName(string moduleFullName) {
+                var index = moduleFullName.LastIndexOf(".");
+                if (index < 0) {
+                    return moduleFullName;
+                }
+
+                return moduleFullName.Substring(index + 1);
+            }
```

Comment:
I think it is just `moduleFullName.Substring(moduleFullName.LastIndexOf(".")+ 1)`... since it returns -1. Or
 `return index < 0 ? moduleFullName : moduleFullName.Substring(index + 1);

Reply:
sure